### PR TITLE
Fix Negotiate/SPNEGO Kerberos to NTLM fallback on Linux

### PR DESCRIPTION
**Pull request by davidsh**
This PR adds support to our Linux gssapi layer so that we properly handle the Negotiate
protocol (SPNEGO) being able to to fall back from Kerberos to NTLM.

Windows Negotiate SSPI implements the SPNEGO protocol. However, it has built-in error
handling for dealing with quick fallback from Kerberos to NTLM in error cases such as where
the client is not on a domain/Kerberos realm, the client is using a local credential, the
client is talking to a non-domain server, etc.

However, the Linux implementation of SPNEGO doesn't handle those error cases very well.
So, I added the proper retry logic when generating the initial context token. Now, we will
be able to fall back from SPNEGO-wrapped Kerberos to SPNEGO-wrapped NTLM protocol blobs.

A similar bug exists in libcurl which is why Curl and CurlHandler are unable to connect to
Negotiate servers when not part of the Kerberos realm. I am considering proposing my fixes
to the libcurl repo as well.

I added more tests to support this PR and did manual testing. I added some additional
EventSource tracing as well.

Contributes to #34878

### CODE REVIEW

File: src/Native/Unix/System.Net.Security.Native/pal_gssapi.c [hunk 1]
```diff
@@ -144,25 +144,27 @@ NetSecurityNative_ImportUserName(uint32_t* minorStatus, char* inputName, uint32_
     return gss_import_name(minorStatus, &inputNameBuffer, GSS_C_NT_USER_NAME, outputName);
 }
 
-uint32_t NetSecurityNative_ImportPrincipalName(uint32_t* minorStatus,
+uint32_t NetSecurityNative_ImportTargetName(uint32_t* minorStatus,
                                                char* inputName,
                                                uint32_t inputNameLen,
+                                               uint32_t isNtlmTarget,
                                                GssName** outputName)
```
Comment (stephentoub) [inline src/Native/Unix/System.Net.Security.Native/pal_gssapi.c]: Nit: the indenting was previously aligned

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Authentication.cs [hunk 2]
```diff
@@ -511,6 +511,10 @@ public static IEnumerable<object[]> ServerUsesWindowsAuthentication_MemberData()
             yield return new object[] { $"http://{server}/test/auth/ntlm/{authEndPoint}", false };
             yield return new object[] { $"https://{server}/test/auth/ntlm/{authEndPoint}", false };
 
+            // Curlhandler (due to libcurl bug) cannot do Negotiate (SPNEGO) Kerberos to NTM fallback.
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Authentication.cs]: Nit: NTM => NTLM

File: src/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs [hunk 3]
```diff
@@ -38,7 +53,8 @@ public SafeDeleteNegoContext(SafeFreeNegoCredentials credential, string targetNa
             Debug.Assert((null != credential), "Null credential in SafeDeleteNegoContext");
             try
             {
-                _targetName = SafeGssNameHandle.CreatePrincipal(targetName);
+                _targetNameKerberos = SafeGssNameHandle.CreateTarget(targetName, false);
+                _targetNameNtlm = SafeGssNameHandle.CreateTarget(targetName, true);
```
Comment (stephentoub) [inline src/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs]: Nit: can you name the false/true arguments so it's clear what they are?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Authentication.cs (hunk 2 above)
Comment (davidsh) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Authentication.cs]: Fixed

File: src/Native/Unix/System.Net.Security.Native/pal_gssapi.c (hunk 1 above)
Comment (davidsh) [reply src/Native/Unix/System.Net.Security.Native/pal_gssapi.c]: Fixed

File: src/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs (hunk 3 above)
Comment (davidsh) [reply src/Common/src/System/Net/Security/Unix/SafeDeleteNegoContext.cs]: Fixed

**Comment by vcsjones** [comment]
@davidsh this is going to sound like a weird suggestion but try doing one pipeline trigger per comment.

**Comment by vcsjones** [comment]
Ugh. That's frustrating. I learned from [this PR](https://github.com/dotnet/corefx/pull/35322#issuecomment-463836383) that the pipelines trigger considers everything after `azp` to be part of the pipeline you are trying to trigger _including new lines_.

**Comment by davidsh** [comment]
@vcsjones Thx for the suggestion!  I'm going to be talking to the Infrastructure team about this to make it better.

**Comment by safern** [comment]
@vcsjones please don't forget to ping me if you have any feedback to bring it over to azure devops folks to make this experience better

**Comment by hacst** [comment]
Are there plans to backport this into the 2.2 (or 2.1) series?

**Comment by davidsh** [comment]
>Are there plans to backport this into the 2.2 (or 2.1) series?

This is a fix among several that I've already made addressing Enterprise Authentication scenarios.  We are considering backporting them as a group to 2.1/2.2.  But the decision hasn't been made yet.  We are waiting to finish a few more fixes in master branch (.NET Core 3.0) before considering this.

@karelz

**Comment by karelz** [comment]
For porting to 2.1/2.2 we would need a very good and strong business case - number of customers who are blocked without the ports and who cannot upgrade to 3.0 from good reasons.

**Pull request merged**